code/go/0chain.net/smartcontract/storagesc/challenge.go

```diff
@@ -340,9 +350,90 @@ func (sc *StorageSmartContract) blobberPenalty(t *transaction.Transaction,
 	return
 }
 
+var challengeMu = &sync.Mutex{}
+
+func (sc *StorageSmartContract) clearExpiredChallenges(t *transaction.Transaction, balances c_state.StateContextI) error {
+
+	challengeMu.Lock()
+	defer challengeMu.Unlock()
+
+	blobbers, err := sc.getBlobbersList(balances)
+	if err != nil {
+		return err
+	}
+
+	for _, blobber := range blobbers.Nodes.copy() {
+		blobberChallengeObj, err := sc.getBlobberChallenge(blobber.ID, balances)
+		if err != nil {
+			return err
+		}
+		var freshChallenges []*StorageChallenge
+		for _, challenge := range blobberChallengeObj.Challenges {
+
+			var alloc *StorageAllocation
+			alloc, err = sc.getAllocation(challenge.AllocationID, balances)
+			if err != nil {
+				return err
+			}
+
+			details, ok := alloc.BlobberMap[t.ClientID]
+			if !ok {
+				return common.NewError("clear_expired_challenges",
+					"Blobber is not part of the allocation")
+			}
+
+			cct := toSeconds(details.Terms.ChallengeCompletionTime)
+			fresh := challenge.Created+cct >= t.CreationDate
+
+			if !fresh {
+				// time of previous complete challenge (not the current one)
+				// or allocation start time if no challenges
+				var prev = alloc.StartTime
+				if last := blobberChallengeObj.LatestCompletedChallenge; last != nil {
+					prev = last.Created
```

Comment:
@guruhubb I am not sure we can handle it like this, seems like processing of challenges need to happen in a specific order - to properly calculate the reward/penalty amount. I cant find any documentation on this except for comments. This seems to be an architecture issue. Since these challenges would have been roughly created at the same time the penalty will be very low for timeouts, not sure this is intended.

Reply:
Have you looked at https://github.com/0chain/0chain/tree/master/code/go/0chain.net/smartcontract/storagesc regarding payments. The blobber gets paid if they pass challenges, otherwise, after paying the validator, it goes back to client along with slash of their stake. When the chain generates a challenge, the SC assigns random validators. The blobbers periodically check to see if there are challenges and then they send their response to the assigned validators.

Reply:
@Sriep can you look into this PR since you're very familiar with challenge rewards, etc